Objects/dictobject.c

```diff
@@ -1118,6 +1160,262 @@ _Py_dict_lookup(PyDictObject *mp, PyObject *key, Py_hash_t hash, PyObject **valu
     return ix;
 }
 
+static inline void
+ensure_shared_on_read(PyDictObject *mp)
+{
+#ifdef Py_GIL_DISABLED
+    if (!_Py_IsOwnedByCurrentThread((PyObject *)mp) && !IS_DICT_SHARED(mp)) {
+        // We are accessing the dict from another thread then owns
+        // it and we haven't marked it as shared which will ensure
+        // that when we re-size ma_keys or ma_values that we will
+        // free using QSBR.  We need to lock the dictionary to
+        // contend with writes from the owning thread, mark it as
+        // shared, and then we can continue with lock-free reads.
```

Comment:
This paragraph could use some editing for grammar and clarity. Maybe something like:

```
// The first time we access a dict from a non-owning thread we mark it as shared. This
// ensures that a concurrent resize operation will delay freeing the old keys or values using
// QSBR, which is necessary to safely allow concurrent reads without locking...
```